src/v/datalake/proto_to_arrow_struct.h

```diff
@@ -0,0 +1,135 @@
+#pragma once
+
+#include "datalake/proto_to_arrow_interface.h"
+#include "datalake/proto_to_arrow_scalar.h"
+
+#include <arrow/api.h>
+#include <arrow/type.h>
+#include <google/protobuf/message.h>
+
+#include <memory>
+#include <stdexcept>
+
+namespace datalake::detail {
+
+class proto_to_arrow_struct : public proto_to_arrow_interface {
+public:
+    explicit proto_to_arrow_struct(
+      const google::protobuf::Descriptor* message_descriptor) {
+        using namespace detail;
+        namespace pb = google::protobuf;
+
+        // Set up child arrays
+        for (int field_idx = 0; field_idx < message_descriptor->field_count();
+             field_idx++) {
+            auto desc = message_descriptor->field(field_idx);
+            if (desc->cpp_type() == pb::FieldDescriptor::CPPTYPE_INT32) {
```

Comment:
could be it's own series of tests. the code has both sum times (status<>) and exceptions. wondeirng if sumtypes should be enough to represent the paths.